File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp

// TC- O(V+E)
// SC- O(V+E)
class Solution {
private:
    vector<int> order;
    bool hascycle(vector<vector<int>>& adjlist, vector<int>& state, int i){
        if(state[i]==1)
            return true;
        state[i]=1;
        for(auto neigh:adjlist[i]){
            if(state[neigh]!=2){
                if(hascycle(adjlist, state, neigh))
                    return true;
            }
        }
        state[i]=2;
        order.push_back(i);
        return false;
    }
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        //if there is a cycle, you can't finish. else you can
        // create adj list to represent problem as graph
        vector<vector<int>> adjlist(numCourses);
        for(int p=0; p<prerequisites.size(); p++)
            adjlist[prerequisites[p][1]].push_back(prerequisites[p][0]);

        // for it to be a cycle - it should revisit an ancestor in current recursion stack
        // create state array to track visited- 0 is not visited, 1 is visiting, 2 is visited
        vector<int> state(numCourses, 0); 

        // start dfs for every course that is not visited
        for(int i=0; i<numCourses; i++){
            if(state[i]!=2){
                if(hascycle(adjlist, state, i))
                    return {};
            }
        }
        reverse(order.begin(), order.end());
        return order;
    }
};

File: 0210-course-schedule-ii/0210-course-schedule-ii.py
# TC- O(p+n)
# SC- O(p+n)

class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj_mat = defaultdict(list)
        for prereq in prerequisites:
            adj_mat[prereq[1]].append(prereq[0])
        
        state= [0]*numCourses
        order = []
        
        def has_cycle(i):
            if state[i]==2:
                return False
            if state[i]==1:
                return True

            state[i]=1
            for c in adj_mat[i]:
                if has_cycle(c):
                    return True
            state[i]=2
            order.append(i)
            return False

        for i in range(numCourses):
            if state[i]==0:
                if has_cycle(i):
                    return []
        return order[::-1]